shell: optimize resources usage

jakub-uC opened the pull request:
Added Kconfig feature to deactivate autocompletion with the Tab key.
Added `IS_CONFIG` macro to remove dead code when a particular shell feature is not used.
Updated prj_minimal config file.
Moved shell `help` command to be always available for the user.
Updated documentation with memory usage by shell features.

Commit:
shell: move help command to shell.c

The help command is needed to list all available commands
when it is not possible to use the tab key.
Previously when build-in commands were deactivated command
help was not compiled as well.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
shell: history feature optimization

When history feature is not compiled in it makes no sense to
trim the command. In addition when history feature is not active the
shell will not call function history_put.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
shell: select command feature optimalization

Added IS_ENABLED(CONFIG_SHELL_CMDS_SELECT) to remove a dead code.
Added static inline function that returns pointer to the selected
command.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
shell: tab feature optimization

Add options to Kconfig that deactivate the Tab button and
autocompletion feature.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
shell: example: prj_minimal config update

Deactivate all features that are not absolutely needed to run
the shell.

Used CONFIG_CBPRINTF_NANO library to save 1020B.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

pabigot reviewed (changes requested):
This looks pretty good, thanks.

I'd like to see a sentence or two in `doc/reference/shell/index.rst` with some words about how all those neat features listed cost ROM and RAM, and where to find the `minimal.cnf` file that shows how to get a minimal version that can be tuned based on need.

It'd also be very nice if every Kconfig option related to the subsystem could be listed there (with links to the detailed documentation), but I don't think there's infrastructure that supports that, and there may be too many for a hand-maintained list to stay correct.  (That list should really have two parts: the `CONFIG_SHELL_FOO` entries and the `CONFIG_FOO_SHELL` entries, so it's clear what affects the shell, and what shell helpers are available.)  But I'm not asking for those in this PR.

Domino-Jachas commented:
Sure, I will update the documentation accordingly.

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -610,7 +610,7 @@ static int execute(const struct shell *shell)
 
 	/* Parent present means we are in select mode. */
-	if (parent != NULL) {
+	if (IS_ENABLED(SHELL_CMDS_SELECT) && (parent != NULL)) {
```
Should be `CONFIG_SHELL_CMDS_SELECT` instead?

mniestroj replied:
Additionally what do you think about introducing:
```
static inline const struct shell_static_entry *
shell_selected_cmd(const struct shell *shell)
{
	if (IS_ENABLED(CONFIG_SHELL_CMDS_SELECT)) {
		return shell->ctx->selected_cmd;
	}

	return NULL;
}
```
so that you can then convert
```
	const struct shell_static_entry *parent = shell->ctx->selected_cmd;
```
to
```
	const struct shell_static_entry *parent = shell_selected_cmd(shell);
```
and leave above `if (parent != NULL) {` untouched?

Seems like using `shell_selected_cmd(shell)`  instead of `shell->ctx->selected_cmd` in other places can save another 16B. Not much, but at least some additional `IS_ENABLED(CONFIG_SHELL_CMDS_SELECT)` can be dropped in code. Look [here](https://github.com/mniestroj/zephyr/commit/05719d8cf12ac9c7e85a8810186f7101a65fad01).

jakub-uC replied:
good point! I will update it.

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -734,7 +734,7 @@ static int execute(const struct shell *shell)
 				      CONFIG_SHELL_ARGC_MAX);
 
-		if (shell->ctx->selected_cmd) {
+		if (IS_ENABLED(SHELL_CMDS_SELECT) && shell->ctx->selected_cmd) {
```
Should be `CONFIG_SHELL_CMDS_SELECT` instead?

jakub-uC replied:
yes! good catch

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -762,6 +793,10 @@ static void tab_handle(const struct shell *shell)
 	size_t argc;
 	size_t cnt;
 
+	if (!IS_ENABLED(CONFIG_SHELL_TAB)) {
```
I guess this place is not needed, because `tab_handle` is called only in:
```
				if (flag_echo_get(shell) &&
				    IS_ENABLED(CONFIG_SHELL_TAB)) {
					/* If the Tab key is pressed, "history
					 * mode" must be terminated because
					 * tab and history handlers are sharing
					 * the same array: temp_buff.
					 */
					flag_history_exit_set(shell, true);
					tab_handle(shell);
				}
```

jakub-uC replied:
I've just tested it. Keeping it saves 16Bytes. The compiler is configured for size optimization.

mniestroj replied:
This is weird. I've checked that quickly now and I see no code size difference when removed `if (!IS_ENABLED(CONFIG_SHELL_TAB)) {` in `tab_handle`. Also compiled with size optimizations.

jakub-uC replied:
There shall be no `!` in the if

mniestroj replied:
What do you mean? There is `!` in the source code line we are commenting here.

jakub-uC replied:
I am referring to your comment:

` if (!IS_ENABLED(CONFIG_SHELL_TAB)) {`

It has `!` while it shall not.

mniestroj replied:
Just to be clear, I was referring to if-statement in body of `tab_handle()`. It has `!` in your proposed changes and it should.

And I was proposing to remove
```
	if (!IS_ENABLED(CONFIG_SHELL_TAB)) {
		return;
	}
```
from `tab_handle()`.

If you were talking about
```
if (flag_echo_get(shell) &&
    IS_ENABLED(CONFIG_SHELL_TAB)) {
```
in `state_collect()`, then I have no objections there. It should be like it is.

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -395,6 +418,10 @@ static void tab_options_print(const struct shell *shell,
 	size_t str_len = shell_strlen(str);
 	size_t idx = first;
 
+	if (!IS_ENABLED(CONFIG_SHELL_TAB)) {
```
Shoudn't be needed, because already optimized `tab_handle` calls that.

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -303,11 +311,16 @@ static void find_completion_candidates(const struct shell *shell,
+	size_t incompl_cmd_len;
 	size_t idx = 0;
 
+	if (!IS_ENABLED(CONFIG_SHELL_TAB)) {
```
Shoudn't be needed, because already optimized `tab_handle` calls that.

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -233,6 +237,10 @@ static bool tab_prepare(const struct shell *shell,
 			size_t *complete_arg_idx,
 			struct shell_static_entry *d_entry)
 {
+	if (!IS_ENABLED(CONFIG_SHELL_TAB)) {
```
Shoudn't be needed, because already optimized `tab_handle` calls that.

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -109,6 +109,10 @@ static void tab_item_print(const struct shell *shell, const char *option,
 	uint16_t columns;
 	uint16_t diff;
 
+	if (!IS_ENABLED(CONFIG_SHELL_TAB)) {
```
Shoudn't be needed, because already optimized `tab_options_print` calls that.

mniestroj commented:
Also change `s/aptions/options` in `shell: tab feature optimization` commit.

jakub-uC commented on subsys/shell/shell.c:
```diff
@@ -599,10 +599,11 @@ static int execute(const struct shell *shell)
-		    shell->ctx->cmd_buff_len);
+	if (IS_ENABLED(CONFIG_SHELL_HISTORY)) {
+		shell_cmd_trim(shell);
+		history_put(shell, shell->ctx->cmd_buff,
```
`history_put` is a local function for shell.c file. I can of course rename it but there are many other local functions in this file that do not have `shell_` prefix. Shall I update all of them?

mniestroj replied:
My point was that `history_put` differs from `shell_history_put` only by a `IS_ENABLED(CONFIG_SHELL_HISTORY)` check inside. In that case there would be no point of keeping `history_put()` anymore, because `shell_history_put()` could be used with the same effect. However looking closer at it, it seems that those local functions tend to improve readability. So maybe lets stick with `history_put()` for now.

pabigot commented:
Looking at the problems getting CI to pass, I think it might make sense to modify `tests/shell/testcase.yaml` to have alternatives to build-only test each of the enable/disable options independently.   Something like https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/unit/cbprintf/testcase.yaml but add `build_only: true`.

Commit:
shell: documentation update

Updated the documantation with newly added configuration features.
Added information where to find minimal shell config file.
Added information how to activate particular features.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

nordic-krch commented on doc/reference/shell/index.rst:
```diff
@@ -542,6 +542,131 @@ are :c:macro:`SHELL_DEFINE` arguments.
+***************
+
+The following is an indication of the consumption of Flash and RAM by each
+shell module. These modules can be deactivated as needed to save CPU resources.
```
you should mention which platform was used. Footprint will vary.

pabigot commented on doc/reference/shell/index.rst:
```diff
@@ -542,6 +542,132 @@ are :c:macro:`SHELL_DEFINE` arguments.
++------------------+-----------+---------+--------------------------------------------------------------------+
+| Feature Name      | Flash [B] | RAM [B] |  Comment                                                          |
++==================+===========+=========+====================================================================+
+|  Shell           | 7012      | 1944    | | Shell module with very basic functionality.                      |
```
Tables like this are going to get out of date really fast, and unless there's a way to automatically generate them I don't think they belong in the documentation.

I personally wouldn't keep this as documentation, but I would put each feature set into the shell test, to verify builds continue to work with the different combinations.  Just build, no need to have special tests for the different combinations---though there should be tests to confirm the behavior resulting from disabling a feature, where that's feasible.

The artifacts resulting could be used to calculate the impact of the changes.  If you want you could document the observed relative changes as comments in the testcase.yaml file so people can verify them over time without making people expect them to be accurate.

jakub-uC replied:
Fair point.
I can rework tests and integrate them with the documentation but I think as a separate PR. This looks like a major piece of work for me.
What do you think?

pabigot replied:
I personally think the various configurations should be build-tested to ensure they continue to work as intended, and that should be done here.  If nobody agrees I'll yield.

But I don't think it's worth integrating the analysis of size impact into the user documentation, certainly not in this PR.

pabigot replied:
Ah; for the documentation part I was only looking for something like this:
```
diff --git a/doc/reference/shell/index.rst b/doc/reference/shell/index.rst
index 312b01316b..1d89cd3711 100644
--- a/doc/reference/shell/index.rst
+++ b/doc/reference/shell/index.rst
@@ -32,6 +32,14 @@ interaction is required. This module is a Unix-like shell with these features:
 * Support for meta keys.
 * Kconfig configuration to optimize memory usage.
 
+Some of these features have a significant impact on RAM and flash usage,
+but many can be disabled when not needed.  A configuration that should
+produce the minimum useful feature set is in
+:zephyr_file:`samples/subsys/shell/shell_module/prj_minimal.conf`:
+
+.. literalinclude:: ../../../samples/subsys/shell_module/prj_minimal.conf
+   :langage: text                  
+
 The module can be connected to any transport for command input and output.
 At this point, the following transport layers are implemented:
```

(Until I looked I wasn't aware of the ability to include the config file into the documentation.  The content above is based on in-tree material but I haven't tested it to see what it looks like.  Paths and rST syntax may be incorrect.  It may also be better placed somewhere after the overall description.)

The desire to have the test cases that you documented build-tested was based on https://github.com/zephyrproject-rtos/zephyr/pull/30116#issuecomment-730326468 and taking advantage of the distinct configurations you identified in your space analysis update.  I think retaining the existing minimal configuration, while adding configurations to individually disable the feature sets you identified, would be helpful both to users and for ensuring the flags continue to work (there are no missed dependencies).  You should be able to use `extra_args: OVERLAY_CONFIG="disable_foo.conf"` in `testcase.yaml` to separate the build tests.

Commit:
shell: help feature improvement

Shell help printing function can now fully support
CONFIG_CBPRINTF_NANO.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
tests: add more configurations for shell tests

Add configurations to test shell feature sets. It will be helpful
both to users and for ensuring the flags continue to work.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

jakub-uC commented:
@pabigot @nordic-krch @mniestroj I think I've addressed all issues.

pabigot reviewed (commented):
Looks pretty good, just a couple tweaks.

jakub-uC commented on doc/reference/shell/index.rst:
```diff
@@ -21,17 +21,22 @@ interaction is required. This module is a Unix-like shell with these features:
 * Support for ANSI escape codes: ``VT100`` and ``ESC[n~`` for cursor control
   and color printing.
-* Support for multiline commands.
+* Support for multiline commands edition.
```
Yes, I meant that. Thank you, I will update the text.

jakub-uC commented on doc/reference/shell/index.rst:
```diff
@@ -432,6 +476,36 @@ The shell module supports the following meta keys:
+new message is enqueued. Use the ``shell stats show`` command to retrieve
+number of log messages dropped by the shell instance. Log queue size and timeout
+are :c:macro:`SHELL_DEFINE` arguments.
+This feature is activated by: :option:`CONFIG_SHELL_LOG_BACKEND` set to ``y``.
```
will fix

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -610,7 +620,7 @@ static int execute(const struct shell *shell)
 
 	/* Parent present means we are in select mode. */
-	if (parent != NULL) {
+	if (IS_ENABLED(CONFIG_SHELL_CMDS_SELECT) && (parent != NULL)) {
```
I don't see any code size improvement in this case, because `gcc` already knows that `parent = selected_cmd_get(shell)` is going to be `NULL` when `CONFIG_SHELL_CMDS_SELECT=n`.

jakub-uC replied:
fixed

mniestroj commented on subsys/shell/shell.c:
```diff
@@ -730,11 +740,12 @@ static int execute(const struct shell *shell)
 
-		if (shell->ctx->selected_cmd) {
+		if (IS_ENABLED(CONFIG_SHELL_CMDS_SELECT) &&
+			       selected_cmd_get(shell)) {
```
Same here, no code size improvement, because `selected_cmd_get()` returns `false` when `CONFIG_SHELL_CMDS_SELECT=n`.

jakub-uC replied:
fixed

mniestroj reviewed (commented):
Just minor things, that don't seem to make any improvement.
